lxd/instance/drivers/driver_common.go

```diff
@@ -503,3 +503,21 @@ func (d *common) updateProgress(progress string) {
 		d.op.UpdateMetadata(meta)
 	}
 }
+
+// lifecycle is used to send a lifecycle event with some instance context.
+func (d *common) lifecycle(action string, ctx map[string]interface{}) error {
+	prefix := "instance"
+	url := fmt.Sprintf("/1.0/instances/%s", d.name)
+
+	if d.snapshot {
+		name, snapName, _ := shared.InstanceGetParentAndSnapshotName(d.name)
+		url = fmt.Sprintf("/1.0/instances/%s/snapshots/%s", name, snapName)
```

Comment:
@stgraber we should probably start using `url.PathEscape()` on variables we are encoding into URL paths, for belt and braces protection against generating invalid URLs.